dm/worker/subtask_test.go

```diff
@@ -0,0 +1,434 @@
+	c.Assert(st.Stage(), Equals, pb.Stage_Paused)
+
+	// restore from pausing
+	c.Assert(st.Resume(), IsNil)
+	c.Assert(st.CurrUnit(), Equals, mockLoader)
+	c.Assert(st.Result(), IsNil)
+	c.Assert(st.Stage(), Equals, pb.Stage_Running)
+
+	// update in running
+	c.Assert(st.Update(nil), NotNil)
+	c.Assert(st.CurrUnit(), Equals, mockLoader)
+	c.Assert(st.Result(), IsNil)
+	c.Assert(st.Stage(), Equals, pb.Stage_Running)
+
+	// Pause
+	c.Assert(st.Pause(), IsNil)
+	c.Assert(st.Stage(), Equals, pb.Stage_Paused)
+	c.Assert(st.CurrUnit(), Equals, mockLoader)
+	if st.Result() != nil && (!st.Result().IsCanceled || len(st.Result().Errors) > 0) {
+		c.Fatalf("result %+v is not right after closing", st.Result())
+	}
+
+	// update again
+	c.Assert(st.Update(&config.SubTaskConfig{Name: "updateSubtask"}), IsNil)
+	c.Assert(st.Stage(), Equals, pb.Stage_Paused)
+	c.Assert(st.CurrUnit(), Equals, mockLoader)
+	if st.Result() != nil && (!st.Result().IsCanceled || len(st.Result().Errors) > 0) {
+		c.Fatalf("result %+v is not right after closing", st.Result())
+	}
+
+	// run again
+	c.Assert(st.Resume(), IsNil)
+	c.Assert(st.CurrUnit(), Equals, mockLoader)
+	c.Assert(st.Result(), IsNil)
+	c.Assert(st.Stage(), Equals, pb.Stage_Running)
+
+	// pause again
+	c.Assert(st.Pause(), IsNil)
+	c.Assert(st.Stage(), Equals, pb.Stage_Paused)
+	c.Assert(st.CurrUnit(), Equals, mockLoader)
+	if st.Result() != nil && (!st.Result().IsCanceled || len(st.Result().Errors) > 0) {
+		c.Fatalf("result %+v is not right after closing", st.Result())
+	}
+
+	// run again
+	c.Assert(st.Resume(), IsNil)
+	c.Assert(st.CurrUnit(), Equals, mockLoader)
+	c.Assert(st.Result(), IsNil)
+	c.Assert(st.Stage(), Equals, pb.Stage_Running)
+
+	// finish loader
+	c.Assert(mockLoader.InjectProcessError(context.Background(), nil), IsNil)
+	for i := 0; i < 1000; i++ {
+		if st.Stage() == pb.Stage_Finished {
+			break
+		}
+		time.Sleep(time.Millisecond)
+	}
+	c.Assert(st.CurrUnit(), Equals, mockLoader)
+	c.Assert(st.Stage(), Equals, pb.Stage_Finished)
+	c.Assert(st.Result().Errors, HasLen, 0)
+}
+
+func (t *testSubTask) TestPauseAndResumeSubtask(c *C) {
+	cfg := &config.SubTaskConfig{
+		Name: "testSubtaskScene",
+		Mode: config.ModeFull,
+	}
+
+	st := NewSubTask(cfg)
+	c.Assert(st.Stage(), DeepEquals, pb.Stage_New)
+
+	mockDumper := NewMockUnit(pb.UnitType_Dump)
+	mockLoader := NewMockUnit(pb.UnitType_Load)
+	st.units = []unit.Unit{mockDumper, mockLoader}
+
+	st.Run()
+	c.Assert(st.Stage(), Equals, pb.Stage_Running)
+	c.Assert(st.CurrUnit(), Equals, mockDumper)
+	c.Assert(st.Error(), IsNil)
+	c.Assert(st.Result(), IsNil)
+
+	// pause twice
+	c.Assert(st.Pause(), IsNil)
+	c.Assert(st.Stage(), Equals, pb.Stage_Paused)
+	c.Assert(st.CurrUnit(), Equals, mockDumper)
+	if st.Result() != nil && (!st.Result().IsCanceled || len(st.Result().Errors) > 0) {
+		c.Fatalf("result %+v is not right after closing", st.Result())
+	}
+
+	c.Assert(st.Pause(), NotNil)
```

Comment:
use `ErrorMatches` for non-nil error to match the exact error message?

Reply:
no need now, I would change it